### ISSUE TITLE
Spawning threads with open directory handles causes a crash

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10387]
Migrated from [rt.perl.org#75154](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=75154) (status was 'resolved')

Searchable as RT75154$

**Comment by p5pRT** [comment]
### From alexford@live.com

#### Created by alexford@live.com


Quick summary&#8203;: if you do an opendir\(\) and then spawn threads\, perl will crash\.
I've included a brief test case below\.  The output is something like this&#8203;:"ok 1Free to wrong pool 3221910 not 3b8580\."
\#\!/usr/bin/perluse threads;
do \{    my $thread = async \{ 42 \};    opendir my $dir\, '\.';    $thread\->join;
&nbsp;    print "ok 1\\n";\};
do \{    opendir my $dir\, '\.';    threads&#8203;::join\(async \{ 42 \}\);
&nbsp;    print "ok 2\\n";\};

<details><summary>Perl Info</summary>

```Flags:    category=library    severity=high    module=threads---Site configuration information for perl 5.12.0:
Configured by sshd_server at Sat Apr 10 22:54:35 2010.
Summary of my perl5 (revision 5 version 12 subversion 0) configuration:     Platform:    osname=MSWin32, osvers=5.2, archname=MSWin32-x64-multi-thread    uname=''    config_args='undef'    hint=recommended, useposix=true, d_sigaction=undef    useithreads=define, usemultiplicity=define    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef    use64bitint=define, use64bitall=undef, uselongdouble=undef    usemymalloc=n, bincompat5005=undef  Compiler:    cc='C:/Perl/site/bin/gcc.exe', ccflags ='-DNDEBUG -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -DWIN64 -DCONSERVATIVE -DUSE_SITECUSTOMIZE -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX -DHASATTRIBUTE -fno-strict-aliasing -mms-bitfields',    optimize='-O2',    cppflags='-DWIN32'    ccversion='', gccversion='3.4.5 (mingw-vista special r3)', gccosandvers=''    intsize=4, longsize=4, ptrsize=8, doublesize=8, byteorder=12345678    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=8    ivtype='__int64', ivsize=8, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8    alignbytes=8, prototype=define  Linker and Libraries:    ld='C:\Perl\site\bin\g++.exe', ldflags ='-L"C:\Perl64\lib\CORE"'    libpth=\lib    libs=-lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32 -lbufferoverflowU -lmsvcrt    perllibs=-lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32 -lbufferoverflowU -lmsvcrt    libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl512.lib    gnulibc_version=''  Dynamic Linking:    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '    cccdlflags=' ', lddlflags='-mdll -L"C:\Perl64\lib\CORE"'
Locally applied patches:    ACTIVEPERL_LOCAL_PATCHES_ENTRY
---@INC for perl 5.12.0:    C:/Perl64/site/lib    C:/Perl64/lib    .
---Environment for perl 5.12.0:    HOME (unset)    LANG (unset)    LANGUAGE (unset)    LD_LIBRARY_PATH (unset)    LOGDIR (unset)    PATH=%SystemRoot%\system32\WindowsPowerShell\v1.0\;C:\Perl64\site\bin;C:\Perl64\bin;C:\Perl\site\bin;C:\Perl\bin;C:\Program Files (x86)\Haskell\bin;C:\Program Files (x86)\Haskell Platform\2010.1.0.0\lib\extralibs\bin;C:\Program Files (x86)\Haskell Platform\2010.1.0.0\bin;c:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\Haskell Platform\2010.1.0.0\mingw\bin;C:\Program Files (x86)\QuickTime\QTSystem\;C:\CUDA\bin64;C:\Users\Alex\bin;C:\Users\Alex\AppData\Roaming\cabal\bin;C:\texlive\2009\bin\win32    PERL_BADLANG (unset)    SHELL (unset) 		 	   		  
_________________________________________________________________
The New Busy is not the too busy. Combine all your e-mail accounts with Hotmail.
http://www.windowslive.com/campaign/thenewbusy?tile=multiaccount&ocid=PID28326::T:WLMTAGL:ON:WL:en-US:WM_HMP:042010_4
```
</details>

**Comment by p5pRT** [comment]
### From @tsee

Hi Alex\,

Alexander Ford wrote&#8203;:
> \# New Ticket Created by  Alexander Ford 
> \# Please include the string&#8203;:  \[perl \#75154\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=75154 >

> Quick summary&#8203;: if you do an opendir\(\) and then spawn threads\, perl will crash\.
> I've included a brief test case below\. 
 > The output is something like this&#8203;:
 > "ok 1Free to wrong pool 3221910 not 3b8580\."

> \#\!/usr/bin/perluse threads;
> do \{    my $thread = async \{ 42 \};    opendir my $dir\, '\.';    $thread\->join;
>     print "ok 1\\n";\};
> do \{    opendir my $dir\, '\.';    threads&#8203;::join\(async \{ 42 \}\);
>     print "ok 2\\n";\};

I only had a system perl 5\.10\.0 with a recent threads\.pm from CPAN 
handy\. There\, I get a crash \(double free or corruption\) only after "ok 
2"\. Since it doesn't reach a statement outside the second do\{\} block\, I 
assume it's happening during the closing of the dir handle or 
respectively the freeing of the $dir variable\.

Best regards\,
Steffen

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jdhedden

Steffen\, Alex\,

What version of threads are you using?

&nbsp;    perl \-Mthreads \-e 'print $threads&#8203;::VERSION\, "\\n"'

I tried your test with threads 1\.77 on 5\.8\.8/9\, 5\.10\.0/1\, 5\.12\.0/1 and
blead\, and did not get a failure\.

If you can upgrade to threads 1\.77 and test again\, it may be that this
problem has already been fixed\.

Thanks\,
Jerry

On Tue\, May 18\, 2010 at 05&#8203;:52\, Steffen Mueller \<smueller@&#8203;cpan\.org> wrote&#8203;:
> Hi Alex\,
>
> Alexander Ford wrote&#8203;:
>>
>> \# New Ticket Created by  Alexander Ford \# Please include the string&#8203;:
>>  \[perl \#75154\]
>> \# in the subject line of all future correspondence about this issue\. \#
>> \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=75154 >
>
>> Quick summary&#8203;: if you do an opendir\(\) and then spawn threads\, perl will
>> crash\.
>> I've included a brief test case below\.
>
>> The output is something like this&#8203;:
>> "ok 1Free to wrong pool 3221910 not 3b8580\."
>
>> \#\!/usr/bin/perluse threads;
>> do \{    my $thread = async \{ 42 \};    opendir my $dir\, '\.';
>>  $thread\->join;
>>    print "ok 1\\n";\};
>> do \{    opendir my $dir\, '\.';    threads&#8203;::join\(async \{ 42 \}\);
>>    print "ok 2\\n";\};
>
> I only had a system perl 5\.10\.0 with a recent threads\.pm from CPAN handy\.
> There\, I get a crash \(double free or corruption\) only after "ok 2"\. Since it
> doesn't reach a statement outside the second do\{\} block\, I assume it's
> happening during the closing of the dir handle or respectively the freeing
> of the $dir variable\.
>
> Best regards\,
> Steffen
>

**Comment by p5pRT** [comment]
### From @cpansprout

This is the result of this function in sv\.c&#8203;:

DIR \*
Perl\_dirp\_dup\(pTHX\_ DIR \*const dp\)
\{
&nbsp;     PERL\_UNUSED\_CONTEXT;
&nbsp;     if \(\!dp\)
&nbsp;    return \(DIR\*\)NULL;
&nbsp;     /\* XXX TODO \*/
&nbsp;     return dp;
\}

Unfortunately\, I don’t know how to duplicate a directory handle  
properly\.

The attached patch \(not to be applied; just an example\) makes a copy  
using dup2\, but the resulting dir handles share the same iterator  
\(i\.e\.\, the current behaviour\, but without the crash\)\. The code works  
on Mac OS X\, but I doubt it’s portable\. I don’t know how to make it  
work on other systems\.

I can see three ultimate solutions \(in order of preference\)&#8203;:

a\) Properly duplicate the dir handle so we have a new one with its own  
iterator \(I can’t find a fopendir function anywhere\)
b\) Simple duplicate the dir handle as in my example
c\) Don’t copy dir handles into threads \(return \(DIR\*\)NULL; I tried  
that and it works\)

If b or c happens\, I think this needs to be documented in threads\.pm\.  
We may have to choose between different items in the list at configure\- 
time\, based on OS\.

**Comment by p5pRT** [comment]
### From @cpansprout

<details><summary>Inline Patch</summary>

```diff
--- blead/sv.c	2010-05-21 12:02:12.000000000 -0700
+++ blead-75154/sv.c	2010-05-23 15:47:28.000000000 -0700
@@ -10658,11 +10611,22 @@ Perl_fp_dup(pTHX_ PerlIO *const fp, cons
 DIR *
 Perl_dirp_dup(pTHX_ DIR *const dp)
 {
+    DIR *ret;
+
     PERL_UNUSED_CONTEXT;
     if (!dp)
 	return (DIR*)NULL;
-    /* XXX TODO */
-    return dp;
+
+    /* look for it in the table first */
+    ret = (DIR*)ptr_table_fetch(PL_ptr_table, dp);
+    if (ret)
+	return ret;
+
+    /* create anew and remember what it is */
+    ret = PerlDir_open(".");
+    PerlLIO_dup2(my_dirfd(dp), my_dirfd(ret));
+    ptr_table_store(PL_ptr_table, dp, ret);
+    return ret;
 }
 
 /* duplicate a typeglob */
```
</details>

**Comment by p5pRT** [comment]
### From @tsee

Hi Jerry\,

Jerry D\. Hedden wrote&#8203;:
> What version of threads are you using?
> 
>     perl \-Mthreads \-e 'print $threads&#8203;::VERSION\, "\\n"'
> 
> I tried your test with threads 1\.77 on 5\.8\.8/9\, 5\.10\.0/1\, 5\.12\.0/1 and
> blead\, and did not get a failure\.
> 
> If you can upgrade to threads 1\.77 and test again\, it may be that this
> problem has already been fixed\.

It was threads 1\.75\. Installing threads 1\.77 from CPAN into the ubuntu 
system 5\.10 results in threads 1\.77 being shadowed by the old core 1\.75\. 
My currently handy copy of bleadperl \(static build\, so YMMV\) gives me 
this backtrace&#8203;:

\#0  0x00007ffff6ba04b5 in \*\_\_GI\_raise \(sig=\<value optimized out>\) at 
\.\./nptl/sysdeps/unix/sysv/linux/raise\.c&#8203;:64
\#1  0x00007ffff6ba3f50 in \*\_\_GI\_abort \(\) at abort\.c&#8203;:92
\#2  0x00007ffff6bd91b7 in \_\_libc\_message \(do\_abort=\<value optimized 
out>\, fmt=\<value optimized out>\) at 
\.\./sysdeps/unix/sysv/linux/libc\_fatal\.c&#8203;:189
\#3  0x00007ffff6be32f6 in malloc\_printerr \(action=3\, str=0x7ffff6ca4cf0 
"double free or corruption \(\!prev\)"\, ptr=\<value optimized out>\) at 
malloc\.c&#8203;:6217
\#4  0x00007ffff6be7c6c in \*\_\_GI\_\_\_libc\_free \(mem=\<value optimized out>\) 
at malloc\.c&#8203;:3716
\#5  0x00007ffff6c0d0bd in \_\_closedir \(dirp=0x26fa\) at 
\.\./sysdeps/unix/closedir\.c&#8203;:52
\#6  0x000000000086d1b1 in Perl\_sv\_clear \(my\_perl=0xd96010\, sv=0xd99a00\) 
at sv\.c&#8203;:5842
\#7  0x000000000086e5cc in Perl\_sv\_free2 \(my\_perl=0xd96010\, sv=0xd99a00\) 
at sv\.c&#8203;:6064
\#8  0x0000000000720d67 in Perl\_gp\_free \(my\_perl=0xd96010\, gv=0xe537f8\) 
at gv\.c&#8203;:1644
\#9  0x000000000086d862 in Perl\_sv\_clear \(my\_perl=0xd96010\, sv=0xe537f8\) 
at sv\.c&#8203;:5887
\#10 0x000000000086e5cc in Perl\_sv\_free2 \(my\_perl=0xd96010\, sv=0xe537f8\) 
at sv\.c&#8203;:6064
\#11 0x0000000000886f4f in Perl\_sv\_unref\_flags \(my\_perl=0xd96010\, 
ref=0xdf5b68\, flags=1\) at sv\.c&#8203;:9074
\#12 0x0000000000862cca in Perl\_sv\_force\_normal\_flags \(my\_perl=0xd96010\, 
sv=0xdf5b68\, flags=1\) at sv\.c&#8203;:4578
\#13 0x00000000008ea7e7 in Perl\_leave\_scope \(my\_perl=0xd96010\, base=2\) at 
scope\.c&#8203;:906
\#14 0x00000000008e52f1 in Perl\_pop\_scope \(my\_perl=0xd96010\) at scope\.c&#8203;:110
\#15 0x0000000000827231 in Perl\_pp\_leave \(my\_perl=0xd96010\) at pp\_hot\.c&#8203;:1919
\#16 0x00000000007cfdf1 in Perl\_runops\_debug \(my\_perl=0xd96010\) at 
dump\.c&#8203;:2112
\#17 0x0000000000701957 in S\_run\_body \(my\_perl=0xd96010\, oldscope=1\) at 
perl\.c&#8203;:2312
\#18 0x0000000000700b7f in perl\_run \(my\_perl=0xd96010\) at perl\.c&#8203;:2236
\#19 0x000000000040769d in main \(argc=2\, argv=0x7fffffffe0a8\, 
env=0x7fffffffe0c0\) at perlmain\.c&#8203;:117


Sorry for the delayed reply\!

Cheers\,
Steffen

**Comment by p5pRT** [comment]
### From sinantrop@gmail.com

This is a bug report for perl from sinantrop@&#8203;gmail\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.12\.1\.

This program crashes perl on windows with those messages&#8203;:
Free to wrong pool 18247e8 not 245a60 at 1\.pl line 11\.
Free to wrong pool 1859c50 not 245a60 at 1\.pl line 11\.

sub ListDir
\{
&nbsp;  my \($dir\, $proc\)=@&#8203;\_;

&nbsp;  local \*DH;
&nbsp;  opendir\(DH\,$dir\);
&nbsp;  while\(my $fn=readdir\(DH\)\)
&nbsp;  \{
&nbsp;    &$proc\($fn\, $dir\);
&nbsp;  \}
&nbsp;  closedir\(DH\);
\}

ListDir\("\."\, sub
\{
&nbsp;  my \($fn\, $dir\)=@&#8203;\_;
&nbsp;  print "$dir/$fn\\n";

&nbsp;  fork;
\}\);


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl 5\.12\.1&#8203;:

Configured by SYSTEM at Fri May 14 00&#8203;:24&#8203;:46 2010\.

Summary of my perl5 \(revision 5 version 12 subversion 1\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=MSWin32\, osvers=5\.00\, archname=MSWin32\-x86\-multi\-thread
&nbsp;    uname=''
&nbsp;    config\_args='undef'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cl'\, ccflags ='\-nologo \-GF \-W3 \-MD \-Zi \-DNDEBUG \-O1 \-DWIN32 
\-D\_CONSOLE \-DNO\_STRICT \-DHAVE\_DES\_FCRYPT \-DUSE\_SITECUSTOMIZE 
\-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS \-DUSE\_PERLIO 
\-DPERL\_MSVCRT\_READFIX'\,
&nbsp;    optimize='\-MD \-Zi \-DNDEBUG \-O1'\,
&nbsp;    cppflags='\-DWIN32'
&nbsp;    ccversion='12\.00\.8804'\, gccversion=''\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=8
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='\_\_int64'\, 
lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='link'\, ldflags ='\-nologo \-nodefaultlib \-debug \-opt&#8203;:ref\,icf  
\-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'
&nbsp;    libpth=\\lib
&nbsp;    libs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib  
comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib  
netapi32\.lib uuid\.lib ws2\_32\.lib mpr\.lib winmm\.lib  version\.lib 
odbc32\.lib odbccp32\.lib comctl32\.lib msvcrt\.lib
&nbsp;    perllibs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib 
winspool\.lib  comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib 
oleaut32\.lib  netapi32\.lib uuid\.lib ws2\_32\.lib mpr\.lib winmm\.lib  
version\.lib odbc32\.lib odbccp32\.lib comctl32\.lib msvcrt\.lib
&nbsp;    libc=msvcrt\.lib\, so=dll\, useshrplib=true\, libperl=perl512\.lib
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags='\-dll \-nologo \-nodefaultlib \-debug 
\-opt&#8203;:ref\,icf  \-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'

Locally applied patches&#8203;:
&nbsp;    ACTIVEPERL\_LOCAL\_PATCHES\_ENTRY
&nbsp;    d956618 Make Term&#8203;::ReadLine&#8203;::findConsole fall back to STDIN if 
/dev/tty can't be opened
&nbsp;    321e50c Escape patch strings before embedding them in patchlevel\.h



---
@&#8203;INC for perl 5\.12\.1&#8203;:
&nbsp;    c&#8203;:/Perl/site/lib
&nbsp;    c&#8203;:/Perl/lib
&nbsp;    \.



---
Environment for perl 5\.12\.1&#8203;:
&nbsp;    HOME \(unset\)
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=C&#8203;:\\WINDOWS\\system32;C&#8203;:\\WINDOWS;C&#8203;:\\WINDOWS\\System32\\Wbem;c&#8203;:\\perl\\bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL \(unset\)

**Comment by p5pRT** [comment]
### From @chorny

On Wed Sep 01 10&#8203;:03&#8203;:57 2010\, strop wrote&#8203;:
> This is a bug report for perl from sinantrop@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.12\.1\.
> 
> This program crashes perl on windows with those messages&#8203;:
> Free to wrong pool 18247e8 not 245a60 at 1\.pl line 11\.
> Free to wrong pool 1859c50 not 245a60 at 1\.pl line 11\.
> 
> sub ListDir
> \{
>   my \($dir\, $proc\)=@&#8203;\_;
> 
>   local \*DH;
>   opendir\(DH\,$dir\);
>   while\(my $fn=readdir\(DH\)\)
>   \{
>     &$proc\($fn\, $dir\);
>   \}
>   closedir\(DH\);
> \}
> 
> ListDir\("\."\, sub
> \{
>   my \($fn\, $dir\)=@&#8203;\_;
>   print "$dir/$fn\\n";
> 
>   fork;
> \}\);
> \-\-\-

I'm not sure that it is a good idea to mix dirhandles and fork/threads\.
This program works correctly&#8203;:
sub ListDir
\{
my \($dir\, $proc\)=@&#8203;\_;

opendir\(my $dh\,$dir\);
my @&#8203;files=readdir\($dh\);
closedir\($dh\);
foreach my $fn \(@&#8203;files\)
\{
&nbsp;  $proc\->\($fn\, $dir\);
\}
\}

ListDir\("\."\, sub
\{
my \($fn\, $dir\)=@&#8203;\_;
print "$dir/$fn\\n";

fork;
\}\);

It's result are meaningless though \- for last file $proc will be
executed 2^N times\, where N is numner of files\.

\-\- 
Alexandr Ciornii\, http&#8203;://chorny\.net

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jdhedden

On Wed Sep 01 10&#8203;:03&#8203;:57 2010\, strop wrote&#8203;:
> This is a bug report for perl from sinantrop@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.12\.1\.
> 
> This program crashes perl on windows with those messages&#8203;:
> Free to wrong pool 18247e8 not 245a60 at 1\.pl line 11\.
> Free to wrong pool 1859c50 not 245a60 at 1\.pl line 11\.
> 
> sub ListDir
> \{
>   my \($dir\, $proc\)=@&#8203;\_;
> 
>   local \*DH;
>   opendir\(DH\,$dir\);
>   while\(my $fn=readdir\(DH\)\)
>   \{
>     &$proc\($fn\, $dir\);
>   \}
>   closedir\(DH\);
> \}
> 
> ListDir\("\."\, sub
> \{
>   my \($fn\, $dir\)=@&#8203;\_;
>   print "$dir/$fn\\n";
> 
>   fork;
> \}\);
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=medium
> \-\-\-
> Site configuration information for perl 5\.12\.1&#8203;:
> 
> Configured by SYSTEM at Fri May 14 00&#8203;:24&#8203;:46 2010\.
> 
> Summary of my perl5 \(revision 5 version 12 subversion 1\) configuration&#8203;:
>    
>   Platform&#8203;:
>     osname=MSWin32\, osvers=5\.00\, archname=MSWin32\-x86\-multi\-thread
>     uname=''
>     config\_args='undef'
>     hint=recommended\, useposix=true\, d\_sigaction=undef
>     useithreads=define\, usemultiplicity=define
>     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
>     use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
>     usemymalloc=n\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='cl'\, ccflags ='\-nologo \-GF \-W3 \-MD \-Zi \-DNDEBUG \-O1 \-DWIN32 
> \-D\_CONSOLE \-DNO\_STRICT \-DHAVE\_DES\_FCRYPT \-DUSE\_SITECUSTOMIZE 
> \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS \-DUSE\_PERLIO 
> \-DPERL\_MSVCRT\_READFIX'\,
>     optimize='\-MD \-Zi \-DNDEBUG \-O1'\,
>     cppflags='\-DWIN32'
>     ccversion='12\.00\.8804'\, gccversion=''\, gccosandvers=''
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>     d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=8
>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='\_\_int64'\, 
> lseeksize=8
>     alignbytes=8\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='link'\, ldflags ='\-nologo \-nodefaultlib \-debug \-opt&#8203;:ref\,icf  
> \-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'
>     libpth=\\lib
>     libs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib  
> comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib  
> netapi32\.lib uuid\.lib ws2\_32\.lib mpr\.lib winmm\.lib  version\.lib 
> odbc32\.lib odbccp32\.lib comctl32\.lib msvcrt\.lib
>     perllibs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib 
> winspool\.lib  comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib 
> oleaut32\.lib  netapi32\.lib uuid\.lib ws2\_32\.lib mpr\.lib winmm\.lib  
> version\.lib odbc32\.lib odbccp32\.lib comctl32\.lib msvcrt\.lib
>     libc=msvcrt\.lib\, so=dll\, useshrplib=true\, libperl=perl512\.lib
>     gnulibc\_version=''
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
>     cccdlflags=' '\, lddlflags='\-dll \-nologo \-nodefaultlib \-debug 
> \-opt&#8203;:ref\,icf  \-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'
> 
> Locally applied patches&#8203;:
>     ACTIVEPERL\_LOCAL\_PATCHES\_ENTRY
>     d956618 Make Term&#8203;::ReadLine&#8203;::findConsole fall back to STDIN if 
> /dev/tty can't be opened
>     321e50c Escape patch strings before embedding them in patchlevel\.h
> 
> \-\-\-
> @&#8203;INC for perl 5\.12\.1&#8203;:
>     c&#8203;:/Perl/site/lib
>     c&#8203;:/Perl/lib
>     \.
> 
> \-\-\-
> Environment for perl 5\.12\.1&#8203;:
>     HOME \(unset\)
>     LANG \(unset\)
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>    
PATH=C&#8203;:\\WINDOWS\\system32;C&#8203;:\\WINDOWS;C&#8203;:\\WINDOWS\\System32\\Wbem;c&#8203;:\\perl\\bin
>     PERL\_BADLANG \(unset\)
>     SHELL \(unset\)
>

**Comment by p5pRT** [comment]
### From @jdhedden

\(Oops\.  Sorry about the previous message\.  Hit the wrong button\.\)

This bug report is really a duplicate of \#75154
&nbsp;  http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=75154

fork\(\) under Windows uses threads\.  And the duplication of dir handles
under threads is not implemented properly\.  \(It's a TODO\.\)

On Wed Sep 01 10&#8203;:03&#8203;:57 2010\, strop wrote&#8203;:
> This is a bug report for perl from sinantrop@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.12\.1\.
> 
> This program crashes perl on windows with those messages&#8203;:
> Free to wrong pool 18247e8 not 245a60 at 1\.pl line 11\.
> Free to wrong pool 1859c50 not 245a60 at 1\.pl line 11\.
> 
> sub ListDir
> \{
>   my \($dir\, $proc\)=@&#8203;\_;
> 
>   local \*DH;
>   opendir\(DH\,$dir\);
>   while\(my $fn=readdir\(DH\)\)
>   \{
>     &$proc\($fn\, $dir\);
>   \}
>   closedir\(DH\);
> \}
> 
> ListDir\("\."\, sub
> \{
>   my \($fn\, $dir\)=@&#8203;\_;
>   print "$dir/$fn\\n";
> 
>   fork;
> \}\);
> \-\-\-

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue May 25 01&#8203;:42&#8203;:10 2010\, sprout wrote&#8203;:
> This is the result of this function in sv\.c&#8203;:
> 
> DIR \*
> Perl\_dirp\_dup\(pTHX\_ DIR \*const dp\)
> \{
>      PERL\_UNUSED\_CONTEXT;
>      if \(\!dp\)
>     return \(DIR\*\)NULL;
>      /\* XXX TODO \*/
>      return dp;
> \}
> 
> Unfortunately\, I don’t know how to duplicate a directory handle  
> properly\.
> 
> The attached patch \(not to be applied; just an example\) makes a copy  
> using dup2\, but the resulting dir handles share the same iterator  
> \(i\.e\.\, the current behaviour\, but without the crash\)\. The code works  
> on Mac OS X\, but I doubt it’s portable\. I don’t know how to make it  
> work on other systems\.
> 
> I can see three ultimate solutions \(in order of preference\)&#8203;:
> 
> a\) Properly duplicate the dir handle so we have a new one with its own  
> iterator \(I can’t find a fopendir function anywhere\)
> b\) Simple duplicate the dir handle as in my example
> c\) Don’t copy dir handles into threads \(return \(DIR\*\)NULL; I tried  
> that and it works\)
> 
> If b or c happens\, I think this needs to be documented in threads\.pm\.  
> We may have to choose between different items in the list at configure\- 
> time\, based on OS\.
> 

Attached is a patch to fix this\. On systems with fchdir\, it dwims \(a\)\.
On other systems the dir handle is simply not passed to the thread \(c\)\.

**Comment by p5pRT** [comment]
### From @cpansprout

From&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>

\[perl \#75174\] Clone dir handles

On systems that support fchdir\, use it to clone dir handles\.

On other systems\, at least for now\, don’t give the new thread a copy
of the handle\. This is not ideal\, but better than crashing\.

<details><summary>Inline Patch</summary>

```diff
diff -up blead/MANIFEST blead-75154-dirdup/MANIFEST
--- blead/MANIFEST	2010-09-11 14:19:12.000000000 -0700
+++ blead-75154-dirdup/MANIFEST	2010-09-17 12:49:00.000000000 -0700
@@ -4635,6 +4635,7 @@ t/op/symbolcache.t		See if undef/delete 
 t/op/sysio.t			See if sysread and syswrite work
 t/op/taint.t			See if tainting works
 t/op/threads_create.pl		Ancillary file for t/op/threads.t
+t/op/threads-dirh.t		Test interaction of threads and dir handles
 t/op/threads.t			Misc. tests for perl features with threads
 t/op/tiearray.t			See if tie for arrays works
 t/op/tie_fetch_count.t		See if FETCH is only called once on tied variables
diff -up blead/sv.c blead-75154-dirdup/sv.c
--- blead/sv.c	2010-09-09 12:00:10.000000000 -0700
+++ blead-75154-dirdup/sv.c	2010-09-18 06:21:19.000000000 -0700
@@ -10849,11 +10849,95 @@ Perl_fp_dup(pTHX_ PerlIO *const fp, cons
 DIR *
 Perl_dirp_dup(pTHX_ DIR *const dp)
 {
+#ifdef HAS_FCHDIR
+    DIR *ret;
+    DIR *pwd;
+    register const Direntry_t *dirent;
+    char smallbuf[256];
+    char *name = NULL;
+    STRLEN len;
+    long pos;
+#endif
+
     PERL_UNUSED_CONTEXT;
+
+#ifdef HAS_FCHDIR
     if (!dp)
 	return (DIR*)NULL;
-    /* XXX TODO */
-    return dp;
+    /* look for it in the table first */
+    ret = (DIR*)ptr_table_fetch(PL_ptr_table, dp);
+    if (ret)
+	return ret;
+
+    /* create anew */
+
+    /* open the current directory (so we can switch back) */
+    if (!(pwd = PerlDir_open("."))) return (DIR *)NULL;
+
+    /* chdir to our dir handle and open the present working directory */
+    if (fchdir(my_dirfd(dp)) < 0 || !(ret = PerlDir_open("."))) {
+	PerlDir_close(pwd);
+	return (DIR *)NULL;
+    }
+    /* Now we should have two dir handles pointing to the same dir. */
+
+    /* Be nice to the calling code and chdir back to where we were. */
+    fchdir(my_dirfd(pwd)); /* If this fails, then what? */
+
+    /* We have no need of the pwd handle any more. */
+    PerlDir_close(pwd);
+
+    /* Iterate once through dp, to get the file name at the current posi-
+       tion. Then step back. */
+    pos = PerlDir_tell(dp);
+    if ((dirent = PerlDir_read(dp))) {
+	len = dirent->d_namlen;
+	if (len <= sizeof smallbuf) name = smallbuf;
+	else Newx(name, len, char);
+	Move(dirent->d_name, name, len, char);
+    }
+    PerlDir_seek(dp, pos);
+
+    /* Iterate through the new dir handle, till we find a file with the
+       right name. */
+    if (!dirent) /* just before the end */
+	for(;;) {
+	    pos = PerlDir_tell(ret);
+	    if (PerlDir_read(ret)) continue; /* not there yet */
+	    PerlDir_seek(ret, pos); /* step back */
+	    break;
+	}
+    else {
+	const long pos0 = PerlDir_tell(ret);
+	for(;;) {
+	    pos = PerlDir_tell(ret);
+	    if ((dirent = PerlDir_read(ret))) {
+		if (len == dirent->d_namlen
+		 && memEQ(name, dirent->d_name, len)) {
+		    /* found it */
+		    PerlDir_seek(ret, pos); /* step back */
+		    break;
+		}
+		/* else we are not there yet; keep iterating */
+	    }
+	    else { /* This is not meant to happen. The best we can do is
+	              reset the iterator to the beginning. */
+		PerlDir_seek(ret, pos0);
+		break;
+	    }
+	}
+    }
+
+    if (name && name != smallbuf)
+	Safefree(name);
+
+    /* pop it in the pointer table */
+    ptr_table_store(PL_ptr_table, dp, ret);
+
+    return ret;
+#else
+    return (DIR*)NULL;
+#endif
 }
 
 /* duplicate a typeglob */
diff -up blead/dist/threads/lib/threads.pm blead-75154-dirdup/dist/threads/lib/threads.pm
--- blead/dist/threads/lib/threads.pm	2010-07-07 07:22:10.000000000 -0700
+++ blead-75154-dirdup/dist/threads/lib/threads.pm	2010-09-17 01:12:59.000000000 -0700
@@ -1005,6 +1005,16 @@ mutexes that are needed to control funct
 For this reason, the use of C<END> blocks in threads is B<strongly>
 discouraged.
 
+=item Directory handles
+
+In perl 5.14.0 and higher, if your system does not support the C<fchdir> C
+function, directory handles will not be copied to new threads. You can use
+the C<d_fchdir> variable in L<Config.pm|Config> to determine whether your
+system supports it.
+
+In prior perl versions, leaving directory handles open when threads were
+created could result in crashes or memory corruption.
+
 =item Perl Bugs and the CPAN Version of L<threads>
 
 Support for threads extends beyond the code in this module (i.e.,
diff -rNup blead/t/op/threads-dirh.t blead-75154-dirdup/t/op/threads-dirh.t
--- blead/t/op/threads-dirh.t	1969-12-31 16:00:00.000000000 -0800
+++ blead-75154-dirdup/t/op/threads-dirh.t	2010-09-18 06:17:53.000000000 -0700
@@ -0,0 +1,131 @@
+#!perl
+
+# Test interaction of threads and directory handles.
+
+BEGIN {
+     chdir 't' if -d 't';
+     @INC = '../lib';
+     require './test.pl';
+     $| = 1;
+
+     require Config;
+     if (!$Config::Config{useithreads}) {
+        print "1..0 # Skip: no ithreads\n";
+        exit 0;
+     }
+     if ($ENV{PERL_CORE_MINITEST}) {
+       print "1..0 # Skip: no dynamic loading on miniperl, no threads\n";
+       exit 0;
+     }
+
+     plan(6);
+}
+
+use strict;
+use warnings;
+use threads;
+use threads::shared;
+use File::Path;
+use File::Spec::Functions qw 'updir catdir';
+use Cwd 'getcwd';
+
+# Basic sanity check: make sure this does not crash
+fresh_perl_is <<'# this is no comment', 'ok', {}, 'crash when duping dirh';
+   use threads;
+   opendir dir, 'op';
+   async{}->join for 1..2;
+   print "ok";
+# this is no comment
+
+my $dir;
+SKIP: {
+ my $skip = sub {
+   chdir($dir);
+   chdir updir;
+   skip $_[0], 5
+ };
+
+ if(!$Config::Config{d_fchdir}) {
+  $::TODO = 'dir handle cloning currently requires fchdir';
+ }
+
+ my @w :shared; # warnings accumulator
+ local $SIG{__WARN__} = sub { push @w, $_[0] };
+
+ $dir = catdir getcwd(), "thrext$$" . int rand() * 100000;
+
+ rmtree($dir);
+ mkdir($dir);
+
+ # Create a dir structure like this:
+ #   $dir
+ #     |
+ #     `- toberead
+ #            |
+ #            +---- thrit
+ #            |
+ #            +---- rile
+ #            |
+ #            `---- zor
+
+ chdir($dir);
+ mkdir 'toberead';
+ chdir 'toberead';
+ {open my $fh, ">thrit" or &$skip("Cannot create file thrit")}
+ {open my $fh, ">rile" or &$skip("Cannot create file rile")}
+ {open my $fh, ">zor" or &$skip("Cannot create file zor")}
+ chdir updir;
+
+ # Then test that dir iterators are cloned correctly.
+
+ opendir my $toberead, 'toberead';
+ my $start_pos = telldir $toberead;
+ my @first_2 = (scalar readdir $toberead, scalar readdir $toberead);
+ my @from_thread = @{; async { [readdir $toberead ] } ->join };
+ my @from_main = readdir $toberead;
+ is join('-', sort @from_thread), join('-', sort @from_main),
+     'dir iterator is copied from one thread to another';
+ like
+   join('-', "", sort(@first_2, @from_thread), ""),
+   qr/(?<!-rile)-rile-thrit-zor-(?!zor-)/i,
+  'cloned iterator iterates exactly once over everything not already seen';
+
+ seekdir $toberead, $start_pos;
+ readdir $toberead for 1 .. @first_2+@from_thread;
+ is
+   async { readdir $toberead // 'undef' } ->join, 'undef',
+  'cloned dir iterator that points to the end of the directory'
+ ;
+
+ # Make sure the cloning code can handle file names longer than 255 chars
+ SKIP: {
+  chdir 'toberead';
+  open my $fh,
+    ">floccipaucinihilopilification-"
+   . "pneumonoultramicroscopicsilicovolcanoconiosis-"
+   . "lopadotemachoselachogaleokranioleipsanodrimypotrimmatosilphiokarabo"
+   . "melitokatakechymenokichlepikossyphophattoperisteralektryonoptokephal"
+   . "liokinklopeleiolagoiosiraibaphetraganopterygon"
+    or
+     chdir updir,
+     skip("OS does not support long file names (and I mean *long*)", 1);
+  chdir updir;
+  opendir my $dirh, "toberead";
+  my $test_name
+    = "dir iterators can be cloned when the next fn > 255 chars";
+  while() {
+   my $pos = telldir $dirh;
+   my $fn = readdir($dirh);
+   if(!defined $fn) { fail($test_name); last SKIP; }
+   if($fn =~ 'lagoio') { 
+    seekdir $dirh, $pos;
+    last;
+   }
+  }
+  is length async { scalar readdir $dirh } ->join, 257, $test_name;
+ }
+
+ is scalar @w, 0, 'no warnings during all that' or diag @w;
+ chdir updir;
+}
+rmtree($dir);
```
</details>

**Comment by p5pRT** [comment]
### From @jdhedden

This patch is written to rely on dirent\->d\_namlen which doesn't always
exist\.  \(I tried it on Cygwin and got build failures\)\.  I tried
replacing occurrences of "dirent\->d\_namlen" with
"strlen\(dirent\->d\_name\)\+1"\, and it built and tested successfully\.

Also\, there's a build warning about 'len' possibly being
uninitialized\, so initializing it to something like \-1 should work\.

On Sun\, Sep 19\, 2010 at 16&#8203;:10\, Father Chrysostomos via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Tue May 25 01&#8203;:42&#8203;:10 2010\, sprout wrote&#8203;:
>> This is the result of this function in sv\.c&#8203;:
>>
>> DIR \*
>> Perl\_dirp\_dup\(pTHX\_ DIR \*const dp\)
>> \{
>>      PERL\_UNUSED\_CONTEXT;
>>      if \(\!dp\)
>>       return \(DIR\*\)NULL;
>>      /\* XXX TODO \*/
>>      return dp;
>> \}
>>
>> Unfortunately\, I don’t know how to duplicate a directory handle
>> properly\.
>>
>> The attached patch \(not to be applied; just an example\) makes a copy
>> using dup2\, but the resulting dir handles share the same iterator
>> \(i\.e\.\, the current behaviour\, but without the crash\)\. The code works
>> on Mac OS X\, but I doubt it’s portable\. I don’t know how to make it
>> work on other systems\.
>>
>> I can see three ultimate solutions \(in order of preference\)&#8203;:
>>
>> a\) Properly duplicate the dir handle so we have a new one with its own
>> iterator \(I can’t find a fopendir function anywhere\)
>> b\) Simple duplicate the dir handle as in my example
>> c\) Don’t copy dir handles into threads \(return \(DIR\*\)NULL; I tried
>> that and it works\)
>>
>> If b or c happens\, I think this needs to be documented in threads\.pm\.
>> We may have to choose between different items in the list at configure\-
>> time\, based on OS\.
>>
>
> Attached is a patch to fix this\. On systems with fchdir\, it dwims \(a\)\.
> On other systems the dir handle is simply not passed to the thread \(c\)\.
>
>
> From&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
>
> \[perl \#75174\] Clone dir handles
>
> On systems that support fchdir\, use it to clone dir handles\.
>
> On other systems\, at least for now\, don’t give the new thread a copy
> of the handle\. This is not ideal\, but better than crashing\.
>
> diff \-up blead/MANIFEST blead\-75154\-dirdup/MANIFEST
> \-\-\- blead/MANIFEST      2010\-09\-11 14&#8203;:19&#8203;:12\.000000000 \-0700
> \+\+\+ blead\-75154\-dirdup/MANIFEST 2010\-09\-17 12&#8203;:49&#8203;:00\.000000000 \-0700
> @&#8203;@&#8203; \-4635\,6 \+4635\,7 @&#8203;@&#8203; t/op/symbolcache\.t                See if undef/delete
>  t/op/sysio\.t                   See if sysread and syswrite work
>  t/op/taint\.t                   See if tainting works
>  t/op/threads\_create\.pl         Ancillary file for t/op/threads\.t
> \+t/op/threads\-dirh\.t            Test interaction of threads and dir handles
>  t/op/threads\.t                 Misc\. tests for perl features with threads
>  t/op/tiearray\.t                        See if tie for arrays works
>  t/op/tie\_fetch\_count\.t         See if FETCH is only called once on tied variables
> diff \-up blead/sv\.c blead\-75154\-dirdup/sv\.c
> \-\-\- blead/sv\.c  2010\-09\-09 12&#8203;:00&#8203;:10\.000000000 \-0700
> \+\+\+ blead\-75154\-dirdup/sv\.c     2010\-09\-18 06&#8203;:21&#8203;:19\.000000000 \-0700
> @&#8203;@&#8203; \-10849\,11 \+10849\,95 @&#8203;@&#8203; Perl\_fp\_dup\(pTHX\_ PerlIO \*const fp\, cons
>  DIR \*
>  Perl\_dirp\_dup\(pTHX\_ DIR \*const dp\)
>  \{
> \+\#ifdef HAS\_FCHDIR
> \+    DIR \*ret;
> \+    DIR \*pwd;
> \+    register const Direntry\_t \*dirent;
> \+    char smallbuf\[256\];
> \+    char \*name = NULL;
> \+    STRLEN len;
> \+    long pos;
> \+\#endif
> \+
>     PERL\_UNUSED\_CONTEXT;
> \+
> \+\#ifdef HAS\_FCHDIR
>     if \(\!dp\)
>        return \(DIR\*\)NULL;
> \-    /\* XXX TODO \*/
> \-    return dp;
> \+    /\* look for it in the table first \*/
> \+    ret = \(DIR\*\)ptr\_table\_fetch\(PL\_ptr\_table\, dp\);
> \+    if \(ret\)
> \+       return ret;
> \+
> \+    /\* create anew \*/
> \+
> \+    /\* open the current directory \(so we can switch back\) \*/
> \+    if \(\!\(pwd = PerlDir\_open\("\."\)\)\) return \(DIR \*\)NULL;
> \+
> \+    /\* chdir to our dir handle and open the present working directory \*/
> \+    if \(fchdir\(my\_dirfd\(dp\)\) \< 0 || \!\(ret = PerlDir\_open\("\."\)\)\) \{
> \+       PerlDir\_close\(pwd\);
> \+       return \(DIR \*\)NULL;
> \+    \}
> \+    /\* Now we should have two dir handles pointing to the same dir\. \*/
> \+
> \+    /\* Be nice to the calling code and chdir back to where we were\. \*/
> \+    fchdir\(my\_dirfd\(pwd\)\); /\* If this fails\, then what? \*/
> \+
> \+    /\* We have no need of the pwd handle any more\. \*/
> \+    PerlDir\_close\(pwd\);
> \+
> \+    /\* Iterate once through dp\, to get the file name at the current posi\-
> \+       tion\. Then step back\. \*/
> \+    pos = PerlDir\_tell\(dp\);
> \+    if \(\(dirent = PerlDir\_read\(dp\)\)\) \{
> \+       len = dirent\->d\_namlen;
> \+       if \(len \<= sizeof smallbuf\) name = smallbuf;
> \+       else Newx\(name\, len\, char\);
> \+       Move\(dirent\->d\_name\, name\, len\, char\);
> \+    \}
> \+    PerlDir\_seek\(dp\, pos\);
> \+
> \+    /\* Iterate through the new dir handle\, till we find a file with the
> \+       right name\. \*/
> \+    if \(\!dirent\) /\* just before the end \*/
> \+       for\(;;\) \{
> \+           pos = PerlDir\_tell\(ret\);
> \+           if \(PerlDir\_read\(ret\)\) continue; /\* not there yet \*/
> \+           PerlDir\_seek\(ret\, pos\); /\* step back \*/
> \+           break;
> \+       \}
> \+    else \{
> \+       const long pos0 = PerlDir\_tell\(ret\);
> \+       for\(;;\) \{
> \+           pos = PerlDir\_tell\(ret\);
> \+           if \(\(dirent = PerlDir\_read\(ret\)\)\) \{
> \+               if \(len == dirent\->d\_namlen
> \+                && memEQ\(name\, dirent\->d\_name\, len\)\) \{
> \+                   /\* found it \*/
> \+                   PerlDir\_seek\(ret\, pos\); /\* step back \*/
> \+                   break;
> \+               \}
> \+               /\* else we are not there yet; keep iterating \*/
> \+           \}
> \+           else \{ /\* This is not meant to happen\. The best we can do is
> \+                     reset the iterator to the beginning\. \*/
> \+               PerlDir\_seek\(ret\, pos0\);
> \+               break;
> \+           \}
> \+       \}
> \+    \}
> \+
> \+    if \(name && name \!= smallbuf\)
> \+       Safefree\(name\);
> \+
> \+    /\* pop it in the pointer table \*/
> \+    ptr\_table\_store\(PL\_ptr\_table\, dp\, ret\);
> \+
> \+    return ret;
> \+\#else
> \+    return \(DIR\*\)NULL;
> \+\#endif
>  \}
>
>  /\* duplicate a typeglob \*/
> diff \-up blead/dist/threads/lib/threads\.pm blead\-75154\-dirdup/dist/threads/lib/threads\.pm
> \-\-\- blead/dist/threads/lib/threads\.pm   2010\-07\-07 07&#8203;:22&#8203;:10\.000000000 \-0700
> \+\+\+ blead\-75154\-dirdup/dist/threads/lib/threads\.pm      2010\-09\-17 01&#8203;:12&#8203;:59\.000000000 \-0700
> @&#8203;@&#8203; \-1005\,6 \+1005\,16 @&#8203;@&#8203; mutexes that are needed to control funct
>  For this reason\, the use of C\<END> blocks in threads is B\<strongly>
>  discouraged\.
>
> \+=item Directory handles
> \+
> \+In perl 5\.14\.0 and higher\, if your system does not support the C\<fchdir> C
> \+function\, directory handles will not be copied to new threads\. You can use
> \+the C\<d\_fchdir> variable in L\<Config\.pm|Config> to determine whether your
> \+system supports it\.
> \+
> \+In prior perl versions\, leaving directory handles open when threads were
> \+created could result in crashes or memory corruption\.
> \+
>  =item Perl Bugs and the CPAN Version of L\<threads>
>
>  Support for threads extends beyond the code in this module \(i\.e\.\,
> diff \-rNup blead/t/op/threads\-dirh\.t blead\-75154\-dirdup/t/op/threads\-dirh\.t
> \-\-\- blead/t/op/threads\-dirh\.t   1969\-12\-31 16&#8203;:00&#8203;:00\.000000000 \-0800
> \+\+\+ blead\-75154\-dirdup/t/op/threads\-dirh\.t      2010\-09\-18 06&#8203;:17&#8203;:53\.000000000 \-0700
> @&#8203;@&#8203; \-0\,0 \+1\,131 @&#8203;@&#8203;
> \+\#\!perl
> \+
> \+\# Test interaction of threads and directory handles\.
> \+
> \+BEGIN \{
> \+     chdir 't' if \-d 't';
> \+     @&#8203;INC = '\.\./lib';
> \+     require '\./test\.pl';
> \+     $| = 1;
> \+
> \+     require Config;
> \+     if \(\!$Config&#8203;::Config\{useithreads\}\) \{
> \+        print "1\.\.0 \# Skip&#8203;: no ithreads\\n";
> \+        exit 0;
> \+     \}
> \+     if \($ENV\{PERL\_CORE\_MINITEST\}\) \{
> \+       print "1\.\.0 \# Skip&#8203;: no dynamic loading on miniperl\, no threads\\n";
> \+       exit 0;
> \+     \}
> \+
> \+     plan\(6\);
> \+\}
> \+
> \+use strict;
> \+use warnings;
> \+use threads;
> \+use threads&#8203;::shared;
> \+use File&#8203;::Path;
> \+use File&#8203;::Spec&#8203;::Functions qw 'updir catdir';
> \+use Cwd 'getcwd';
> \+
> \+\# Basic sanity check&#8203;: make sure this does not crash
> \+fresh\_perl\_is \<\<'\# this is no comment'\, 'ok'\, \{\}\, 'crash when duping dirh';
> \+   use threads;
> \+   opendir dir\, 'op';
> \+   async\{\}\->join for 1\.\.2;
> \+   print "ok";
> \+\# this is no comment
> \+
> \+my $dir;
> \+SKIP&#8203;: \{
> \+ my $skip = sub \{
> \+   chdir\($dir\);
> \+   chdir updir;
> \+   skip $\_\[0\]\, 5
> \+ \};
> \+
> \+ if\(\!$Config&#8203;::Config\{d\_fchdir\}\) \{
> \+  $&#8203;::TODO = 'dir handle cloning currently requires fchdir';
> \+ \}
> \+
> \+ my @&#8203;w :shared; \# warnings accumulator
> \+ local $SIG\{\_\_WARN\_\_\} = sub \{ push @&#8203;w\, $\_\[0\] \};
> \+
> \+ $dir = catdir getcwd\(\)\, "thrext$$" \. int rand\(\) \* 100000;
> \+
> \+ rmtree\($dir\);
> \+ mkdir\($dir\);
> \+
> \+ \# Create a dir structure like this&#8203;:
> \+ \#   $dir
> \+ \#     |
> \+ \#     \`\- toberead
> \+ \#            |
> \+ \#            \+\-\-\-\- thrit
> \+ \#            |
> \+ \#            \+\-\-\-\- rile
> \+ \#            |
> \+ \#            \`\-\-\-\- zor
> \+
> \+ chdir\($dir\);
> \+ mkdir 'toberead';
> \+ chdir 'toberead';
> \+ \{open my $fh\, ">thrit" or &$skip\("Cannot create file thrit"\)\}
> \+ \{open my $fh\, ">rile" or &$skip\("Cannot create file rile"\)\}
> \+ \{open my $fh\, ">zor" or &$skip\("Cannot create file zor"\)\}
> \+ chdir updir;
> \+
> \+ \# Then test that dir iterators are cloned correctly\.
> \+
> \+ opendir my $toberead\, 'toberead';
> \+ my $start\_pos = telldir $toberead;
> \+ my @&#8203;first\_2 = \(scalar readdir $toberead\, scalar readdir $toberead\);
> \+ my @&#8203;from\_thread = @&#8203;\{; async \{ \[readdir $toberead \] \} \->join \};
> \+ my @&#8203;from\_main = readdir $toberead;
> \+ is join\('\-'\, sort @&#8203;from\_thread\)\, join\('\-'\, sort @&#8203;from\_main\)\,
> \+     'dir iterator is copied from one thread to another';
> \+ like
> \+   join\('\-'\, ""\, sort\(@&#8203;first\_2\, @&#8203;from\_thread\)\, ""\)\,
> \+   qr/\(?\<\!\-rile\)\-rile\-thrit\-zor\-\(?\!zor\-\)/i\,
> \+  'cloned iterator iterates exactly once over everything not already seen';
> \+
> \+ seekdir $toberead\, $start\_pos;
> \+ readdir $toberead for 1 \.\. @&#8203;first\_2\+@&#8203;from\_thread;
> \+ is
> \+   async \{ readdir $toberead // 'undef' \} \->join\, 'undef'\,
> \+  'cloned dir iterator that points to the end of the directory'
> \+ ;
> \+
> \+ \# Make sure the cloning code can handle file names longer than 255 chars
> \+ SKIP&#8203;: \{
> \+  chdir 'toberead';
> \+  open my $fh\,
> \+    ">floccipaucinihilopilification\-"
> \+   \. "pneumonoultramicroscopicsilicovolcanoconiosis\-"
> \+   \. "lopadotemachoselachogaleokranioleipsanodrimypotrimmatosilphiokarabo"
> \+   \. "melitokatakechymenokichlepikossyphophattoperisteralektryonoptokephal"
> \+   \. "liokinklopeleiolagoiosiraibaphetraganopterygon"
> \+    or
> \+     chdir updir\,
> \+     skip\("OS does not support long file names \(and I mean \*long\*\)"\, 1\);
> \+  chdir updir;
> \+  opendir my $dirh\, "toberead";
> \+  my $test\_name
> \+    = "dir iterators can be cloned when the next fn > 255 chars";
> \+  while\(\) \{
> \+   my $pos = telldir $dirh;
> \+   my $fn = readdir\($dirh\);
> \+   if\(\!defined $fn\) \{ fail\($test\_name\); last SKIP; \}
> \+   if\($fn =~ 'lagoio'\) \{
> \+    seekdir $dirh\, $pos;
> \+    last;
> \+   \}
> \+  \}
> \+  is length async \{ scalar readdir $dirh \} \->join\, 257\, $test\_name;
> \+ \}
> \+
> \+ is scalar @&#8203;w\, 0\, 'no warnings during all that' or diag @&#8203;w;
> \+ chdir updir;
> \+\}
> \+rmtree\($dir\);
>
>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Sep 21 10&#8203;:59&#8203;:20 2010\, jdhedden wrote&#8203;:
> This patch is written to rely on dirent\->d\_namlen which doesn't always
> exist\.  \(I tried it on Cygwin and got build failures\)\.  I tried
> replacing occurrences of "dirent\->d\_namlen" with
> "strlen\(dirent\->d\_name\)\+1"\, and it built and tested successfully\.

Why \+1? d\_namlen does not include the null\.

Does this new patch work for you?

**Comment by p5pRT** [comment]
### From @cpansprout

From&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>

\[perl \#75174\] Clone dir handles

On systems that support fchdir\, use it to clone dir handles\.

On other systems\, at least for now\, don’t give the new thread a copy
of the handle\. This is not ideal\, but better than crashing\.

<details><summary>Inline Patch</summary>

```diff
diff -Nup blead/MANIFEST blead-75154-dirdup2/MANIFEST
--- blead/MANIFEST	2010-09-19 16:19:10.000000000 -0700
+++ blead-75154-dirdup2/MANIFEST	2010-09-21 11:16:37.000000000 -0700
@@ -4637,6 +4637,7 @@ t/op/symbolcache.t		See if undef/delete 
 t/op/sysio.t			See if sysread and syswrite work
 t/op/taint.t			See if tainting works
 t/op/threads_create.pl		Ancillary file for t/op/threads.t
+t/op/threads-dirh.t		Test interaction of threads and dir handles
 t/op/threads.t			Misc. tests for perl features with threads
 t/op/tiearray.t			See if tie for arrays works
 t/op/tie_fetch_count.t		See if FETCH is only called once on tied variables
diff -Nup blead/sv.c blead-75154-dirdup2/sv.c
--- blead/sv.c	2010-09-16 07:47:14.000000000 -0700
+++ blead-75154-dirdup2/sv.c	2010-09-21 15:35:57.000000000 -0700
@@ -10730,11 +10730,101 @@ Perl_fp_dup(pTHX_ PerlIO *const fp, cons
 DIR *
 Perl_dirp_dup(pTHX_ DIR *const dp)
 {
+#ifdef HAS_FCHDIR
+    DIR *ret;
+    DIR *pwd;
+    register const Direntry_t *dirent;
+    char smallbuf[256];
+    char *name = NULL;
+    STRLEN len = -1;
+    long pos;
+#endif
+
     PERL_UNUSED_CONTEXT;
+
+#ifdef HAS_FCHDIR
     if (!dp)
 	return (DIR*)NULL;
-    /* XXX TODO */
-    return dp;
+    /* look for it in the table first */
+    ret = (DIR*)ptr_table_fetch(PL_ptr_table, dp);
+    if (ret)
+	return ret;
+
+    /* create anew */
+
+    /* open the current directory (so we can switch back) */
+    if (!(pwd = PerlDir_open("."))) return (DIR *)NULL;
+
+    /* chdir to our dir handle and open the present working directory */
+    if (fchdir(my_dirfd(dp)) < 0 || !(ret = PerlDir_open("."))) {
+	PerlDir_close(pwd);
+	return (DIR *)NULL;
+    }
+    /* Now we should have two dir handles pointing to the same dir. */
+
+    /* Be nice to the calling code and chdir back to where we were. */
+    fchdir(my_dirfd(pwd)); /* If this fails, then what? */
+
+    /* We have no need of the pwd handle any more. */
+    PerlDir_close(pwd);
+
+#ifdef DIRNAMLEN
+# define d_namlen(d) (d)->d_namlen
+#else
+# define d_namlen(d) strlen((d)->d_name)
+#endif
+    /* Iterate once through dp, to get the file name at the current posi-
+       tion. Then step back. */
+    pos = PerlDir_tell(dp);
+    if ((dirent = PerlDir_read(dp))) {
+	len = d_namlen(dirent);
+	if (len <= sizeof smallbuf) name = smallbuf;
+	else Newx(name, len, char);
+	Move(dirent->d_name, name, len, char);
+    }
+    PerlDir_seek(dp, pos);
+
+    /* Iterate through the new dir handle, till we find a file with the
+       right name. */
+    if (!dirent) /* just before the end */
+	for(;;) {
+	    pos = PerlDir_tell(ret);
+	    if (PerlDir_read(ret)) continue; /* not there yet */
+	    PerlDir_seek(ret, pos); /* step back */
+	    break;
+	}
+    else {
+	const long pos0 = PerlDir_tell(ret);
+	for(;;) {
+	    pos = PerlDir_tell(ret);
+	    if ((dirent = PerlDir_read(ret))) {
+		if (len == d_namlen(dirent)
+		 && memEQ(name, dirent->d_name, len)) {
+		    /* found it */
+		    PerlDir_seek(ret, pos); /* step back */
+		    break;
+		}
+		/* else we are not there yet; keep iterating */
+	    }
+	    else { /* This is not meant to happen. The best we can do is
+	              reset the iterator to the beginning. */
+		PerlDir_seek(ret, pos0);
+		break;
+	    }
+	}
+    }
+#undef d_namlen
+
+    if (name && name != smallbuf)
+	Safefree(name);
+
+    /* pop it in the pointer table */
+    ptr_table_store(PL_ptr_table, dp, ret);
+
+    return ret;
+#else
+    return (DIR*)NULL;
+#endif
 }
 
 /* duplicate a typeglob */
diff -Nurp blead/t/op/threads-dirh.t blead-75154-dirdup2/t/op/threads-dirh.t
--- blead/t/op/threads-dirh.t	1969-12-31 16:00:00.000000000 -0800
+++ blead-75154-dirdup2/t/op/threads-dirh.t	2010-09-21 11:16:37.000000000 -0700
@@ -0,0 +1,131 @@
+#!perl
+
+# Test interaction of threads and directory handles.
+
+BEGIN {
+     chdir 't' if -d 't';
+     @INC = '../lib';
+     require './test.pl';
+     $| = 1;
+
+     require Config;
+     if (!$Config::Config{useithreads}) {
+        print "1..0 # Skip: no ithreads\n";
+        exit 0;
+     }
+     if ($ENV{PERL_CORE_MINITEST}) {
+       print "1..0 # Skip: no dynamic loading on miniperl, no threads\n";
+       exit 0;
+     }
+
+     plan(6);
+}
+
+use strict;
+use warnings;
+use threads;
+use threads::shared;
+use File::Path;
+use File::Spec::Functions qw 'updir catdir';
+use Cwd 'getcwd';
+
+# Basic sanity check: make sure this does not crash
+fresh_perl_is <<'# this is no comment', 'ok', {}, 'crash when duping dirh';
+   use threads;
+   opendir dir, 'op';
+   async{}->join for 1..2;
+   print "ok";
+# this is no comment
+
+my $dir;
+SKIP: {
+ my $skip = sub {
+   chdir($dir);
+   chdir updir;
+   skip $_[0], 5
+ };
+
+ if(!$Config::Config{d_fchdir}) {
+  $::TODO = 'dir handle cloning currently requires fchdir';
+ }
+
+ my @w :shared; # warnings accumulator
+ local $SIG{__WARN__} = sub { push @w, $_[0] };
+
+ $dir = catdir getcwd(), "thrext$$" . int rand() * 100000;
+
+ rmtree($dir);
+ mkdir($dir);
+
+ # Create a dir structure like this:
+ #   $dir
+ #     |
+ #     `- toberead
+ #            |
+ #            +---- thrit
+ #            |
+ #            +---- rile
+ #            |
+ #            `---- zor
+
+ chdir($dir);
+ mkdir 'toberead';
+ chdir 'toberead';
+ {open my $fh, ">thrit" or &$skip("Cannot create file thrit")}
+ {open my $fh, ">rile" or &$skip("Cannot create file rile")}
+ {open my $fh, ">zor" or &$skip("Cannot create file zor")}
+ chdir updir;
+
+ # Then test that dir iterators are cloned correctly.
+
+ opendir my $toberead, 'toberead';
+ my $start_pos = telldir $toberead;
+ my @first_2 = (scalar readdir $toberead, scalar readdir $toberead);
+ my @from_thread = @{; async { [readdir $toberead ] } ->join };
+ my @from_main = readdir $toberead;
+ is join('-', sort @from_thread), join('-', sort @from_main),
+     'dir iterator is copied from one thread to another';
+ like
+   join('-', "", sort(@first_2, @from_thread), ""),
+   qr/(?<!-rile)-rile-thrit-zor-(?!zor-)/i,
+  'cloned iterator iterates exactly once over everything not already seen';
+
+ seekdir $toberead, $start_pos;
+ readdir $toberead for 1 .. @first_2+@from_thread;
+ is
+   async { readdir $toberead // 'undef' } ->join, 'undef',
+  'cloned dir iterator that points to the end of the directory'
+ ;
+
+ # Make sure the cloning code can handle file names longer than 255 chars
+ SKIP: {
+  chdir 'toberead';
+  open my $fh,
+    ">floccipaucinihilopilification-"
+   . "pneumonoultramicroscopicsilicovolcanoconiosis-"
+   . "lopadotemachoselachogaleokranioleipsanodrimypotrimmatosilphiokarabo"
+   . "melitokatakechymenokichlepikossyphophattoperisteralektryonoptokephal"
+   . "liokinklopeleiolagoiosiraibaphetraganopterygon"
+    or
+     chdir updir,
+     skip("OS does not support long file names (and I mean *long*)", 1);
+  chdir updir;
+  opendir my $dirh, "toberead";
+  my $test_name
+    = "dir iterators can be cloned when the next fn > 255 chars";
+  while() {
+   my $pos = telldir $dirh;
+   my $fn = readdir($dirh);
+   if(!defined $fn) { fail($test_name); last SKIP; }
+   if($fn =~ 'lagoio') { 
+    seekdir $dirh, $pos;
+    last;
+   }
+  }
+  is length async { scalar readdir $dirh } ->join, 257, $test_name;
+ }
+
+ is scalar @w, 0, 'no warnings during all that' or diag @w;
+ chdir updir;
+}
+rmtree($dir);
```
</details>

**Comment by p5pRT** [comment]
### From @jdhedden

On Sun\, Sep 26\, 2010 at 19&#8203;:42\, Father Chrysostomos via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Tue Sep 21 10&#8203;:59&#8203;:20 2010\, jdhedden wrote&#8203;:
>> This patch is written to rely on dirent\->d\_namlen which doesn't always
>> exist\.  \(I tried it on Cygwin and got build failures\)\.  I tried
>> replacing occurrences of "dirent\->d\_namlen" with
>> "strlen\(dirent\->d\_name\)\+1"\, and it built and tested successfully\.
>
> Why \+1? d\_namlen does not include the null\.
>
> Does this new patch work for you?

Yes\, your new patch compiles and tests successfully\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Sep 27 13&#8203;:01&#8203;:46 2010\, jdhedden wrote&#8203;:
> On Sun\, Sep 26\, 2010 at 19&#8203;:42\, Father Chrysostomos via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > On Tue Sep 21 10&#8203;:59&#8203;:20 2010\, jdhedden wrote&#8203;:
> >> This patch is written to rely on dirent\->d\_namlen which doesn't always
> >> exist\.  \(I tried it on Cygwin and got build failures\)\.  I tried
> >> replacing occurrences of "dirent\->d\_namlen" with
> >> "strlen\(dirent\->d\_name\)\+1"\, and it built and tested successfully\.
> >
> > Why \+1? d\_namlen does not include the null\.
> >
> > Does this new patch work for you?
> 
> Yes\, your new patch compiles and tests successfully\.
> 

Applied as 11a11e\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
